contracts/verifiers/UniversalVerifierMultiQuery.sol

```diff
@@ -0,0 +1,717 @@
+    }
+
+    /**
+     * @dev Gets a specific multi query by ID
+     * @param queryId The ID of the multi query
+     * @return query The query data
+     */
+    function getQuery(uint256 queryId) public view returns (Query memory query) {
+        return _getUniversalVerifierMultiQueryStorage()._queries[queryId];
+    }
+
+    function _checkAuthResponses(Response[] memory responses) internal pure returns (uint256) {
+        uint256 numAuthResponses = 0;
+        uint256 authResponseIndex;
+
+        for (uint256 i = 0; i < responses.length; i++) {
+            if (_getRequestType(responses[i].requestId) == AUTH_REQUEST_TYPE) {
+                authResponseIndex = i;
+                numAuthResponses++;
+            }
+        }
+        require(numAuthResponses == 1, "Exactly 1 auth response is required");
+
+        return authResponseIndex;
+    }
+
+    function _checkRequestsInQuery(uint256 queryId) internal view {
+        UniversalVerifierMultiQueryStorage storage s = _getUniversalVerifierMultiQueryStorage();
+
+        uint256[] memory requestIds = s._queries[queryId].requestIds;
+
+        // check that all the single requests doesn't have group
+        for (uint256 i = 0; i < requestIds.length; i++) {
+            require(
+                s._requests[requestIds[i]].validator.getGroupID(
+                    s._requests[requestIds[i]].params
+                ) == 0,
+                "A single request in this query is a grouped request"
+            );
+        }
+    }
+
+    /**
+     * @dev Submits an array of responses and updates proofs status
+     * @param authResponses The list of auth responses including request ID, proof and metadata for auth requests
+     * @param singleResponses The list of responses including request ID, proof and metadata for single requests
+     * @param groupedResponses The list of responses including request ID, proof and metadata for grouped requests
+     * @param crossChainProofs The list of cross chain proofs from universal resolver (oracle). This
+     * includes identities and global states.
+     */
+    function submitResponse(
+        Response[] memory authResponses,
+        Response[] memory singleResponses,
+        GroupedResponses[] memory groupedResponses,
+        bytes memory crossChainProofs
+    ) public {
+        UniversalVerifierMultiQueryStorage storage $ = _getUniversalVerifierMultiQueryStorage();
+        address sender = _msgSender();
+
+        // 1. Process crossChainProofs
+        $._state.processCrossChainProofs(crossChainProofs);
+
+        // 2. Process auth responses first
+
+        for (uint256 i = 0; i < authResponses.length; i++) {
+            uint256 userIDFromReponse;
+            if (_getRequestType(authResponses[i].requestId) != AUTH_REQUEST_TYPE) {
+                revert("Request ID is not an auth request");
+            }
+            RequestData memory authRequest = $._requests[authResponses[i].requestId];
```

Comment:
Use `RequestData storage authRequest = ... `in such cases. Should be much less gas for storage pointer assignment than allocating full struct copy in memory.